feat(python): expose rust writer as additional engine

ion-elgreco opened the pull request:
# Description
First version of a functional python binding to rust writer. Lots of work is happening around the writer. This one we should merge first: https://github.com/delta-io/delta-rs/pull/1820.

A couple gaps will exist between current Rust writer and pyarrow writer. We will have to solve this in a later PR,:
- Overwrite schema in Rust 
- Replacewhere (partition filter / predicate) overwrite

# Related Issue(s)
- closes https://github.com/delta-io/delta-rs/issues/1861

ion-elgreco commented:
@wjones127 Any early feedback on how to tackle the missing features and whether I am overseeing something would be much appreciated 😄

roeap commented on python/src/lib.rs:
```diff
@@ -1127,6 +1128,71 @@ impl From<&PyAddAction> for Add {
+    configuration: Option<HashMap<String, Option<String>>>,
+    storage_options: Option<HashMap<String, String>>,
+) -> PyResult<()> {
+    let batches = data.0.map(|batch| batch.unwrap()).collect::<Vec<_>>();
```
We should probably try to not fully materialize the RBR into memory. This would entail wrapping the RBR in an `ExecutionPlan`, which can be passed to the `CreateBuilder`. Maybe [`DatasetExec`](https://github.com/apache/arrow-datafusion-python/blob/main/src/dataset_exec.rs) from datafusion-python can provide some inspiration.

ion-elgreco replied:
Maybe we can do that in a second PR then, since I was also doing it like this for MERGE python binding. Then I can improve these two at the same time.

roeap commented on python/src/lib.rs:
```diff
@@ -1127,6 +1128,71 @@ impl From<&PyAddAction> for Add {
+        .write(batches)
+        .with_save_mode(save_mode)
+        .with_overwrite_schema(overwrite_schema)
+        .with_metadata(metadata)
```
This name is a little bit misleading. It does not refer to the metadata action, but rather additional application-specific metadata. Took a quick look at the code, but it seems table configuration it not passed through to the create operation in the "create if not exists" path.

__update__: just saw you mentioned that already :).

roeap commented on crates/deltalake-core/src/operations/write.rs:
```diff
@@ -108,6 +108,8 @@ pub struct WriteBuilder {
     /// RecordBatches to be written into the table
     batches: Option<Vec<RecordBatch>>,
+    /// whether to overwrite the schema
+    overwrite_schema: bool,
```
is this meant for schema evolution? If so, I'd recommend moving that to a follow-up PR as it would likely blow up this PR quite a bit.

wjones127 replied:
+1. I think it's fine if we let that return `NotImplementedError` for now.

ion-elgreco replied:
It was an quick attempt for schema evolution. I was able to write except it didn't write the columns that were not part of the original schema, so I need to dig through the code more.

Ok, let's do this as improvement in another update.

roeap replied:
we also would need to update all read path to always add null columns for columns non-existent in older parquet files. Haven't looked into it, but this would likely require some larger refactoring particularly in the datafusion `DeltaScan`. Saying this we likely need to validate that added columns are always nullable as well.

ion-elgreco replied:
I think that would be schema evolution for also appends.

The PyArrow writer can do schema evolution but only combined with an `overwrite` mode. 

I think that's purely a metadata action then. Would this be doable with the existing deltalake-core crate?

roeap replied:
not sure, we may end up with unreadble tables if we do this... if we replace the whole table this might work.

ion-elgreco replied:
With PyArrow it only works together with overwrite so it should be safe. Is there a way to adjust the commit that's written during a write?

roeap reviewed (commented):
Certainly :). 

Overall I believe this is going in the right direction. one thing that may be a larger piece of work is not collecting the iterator, but rather wrapping it in an execution plan. Maybe we could also live with collect for now to get the wiring right, and leave the execution plan to a follow-up.

wjones127 commented on python/deltalake/writer.py:
To make the diff as clean as possible, could we instead change `write_deltalake` to dispatch to either `_write_deltalake_pyarrow` (current impl) or `_write_deltalake_rust` (new impl)?

ion-elgreco replied:
Yes can do that!

wjones127 commented on python/tests/test_writer.py:
```diff
@@ -29,24 +29,30 @@
 
 
-def test_roundtrip_basic(tmp_path: pathlib.Path, sample_data: pa.Table):
+@pytest.mark.parametrize("engine", ["pyarrow", "rust"])
```
Thanks, this is exactly what I had in mind 😄

ion-elgreco commented:
@roeap @wjones127 it's ready for a full review now. That one failing test will be solved when we merge this PR https://github.com/delta-io/delta-rs/pull/1820, I've also updated the PR description, basically we have two gaps with pyarrow and rust writer.

roeap commented:
While looking through this I realized we may have a bug with the schema overwrite right now.

When we added this, overwrite would always replace the entire table. Since then we added selectively overwriting partitions, in which case I believe we may be corrupting the table, or at least make it unreadable to our readers, since we end up with parquet files with different schemas in the table.

Not sure if the pyarrow Dataset can handle that, but almost certain, our rust readers would not be able to handle that.

Should we include as an additional check, that no partition filters are supplied, if the schema is overwritten?

ion-elgreco commented:
> While looking through this I realized we may have a bug with the schema overwrite right now.
> 
> When we added this, overwrite would always replace the entire table. Since then we added selectively overwriting partitions, in which case I believe we may be corrupting the table, or at least make it unreadable to our readers, since we end up with parquet files with different schemas in the table.
> 
> Not sure if the pyarrow Dataset can handle that, but almost certain, our rust readers would not be able to handle that.
> 
> Should we include as an additional check, that no partition filters are supplied, if the schema is overwritten?

For the rust writer? Because currently only pyarrow is using them.

roeap commented:
> For the rust writer? Because currently only pyarrow is using them.

Yes, and the pyarrow writer may be creating corrupted tables that we cannot read from rust and maybe even python if partition filters are supplied.

ion-elgreco commented:
Closing this in favour of: https://github.com/delta-io/delta-rs/pull/1891
